[FX] Add Interpreter and Transformer

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50420 [FX] Add Interpreter and Transformer**

1. Codify the interpreter pattern that was used, for ex, in [shape prop](https://github.com/pytorch/pytorch/blob/c3b4b2062748ed70e689f9f9ffe670e6fa20a071/torch/fx/experimental/shape_prop.py#L7). This is a pattern users will likely find repeatedly useful so it's good to have it first-class in the API. As seen in the test cases, this makes it incredibly easy to do simple analysis and transforms like swapping out individual nodes

2. Add a small `Transformer` class that is an `Interpreter` but provides a `transform` method that yields a GraphModule from interpreting the code

![image](https://user-images.githubusercontent.com/4685384/105924691-a7b0b700-5ff3-11eb-8de8-7967d6b8ac69.png)
![image](https://user-images.githubusercontent.com/4685384/105924696-abdcd480-5ff3-11eb-8454-a2bc36c3248e.png)


TODO:

- [x] Write docstrings
- [x] Write docs
- [ ] Look for more places this can be used

Differential Revision: [D25880330](https://our.internmc.facebook.com/intern/diff/D25880330)

Commit:
[WIP][FX] Add Interpreter and Transformer

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Add Interpreter and Transformer"




[ghstack-poisoned]

Chillee commented:
I think it might be easier to implement vmap with this.

Commit:
Update on "[WIP][FX] Add Interpreter and Transformer"


1. Codify the interpreter pattern that was used, for ex, in [shape prop](https://github.com/pytorch/pytorch/blob/c3b4b2062748ed70e689f9f9ffe670e6fa20a071/torch/fx/experimental/shape_prop.py#L7). This is a pattern users will likely find repeatedly useful so it's good to have it first-class in the API. As seen in the test cases, this makes it incredibly easy to do simple analysis and transforms like swapping out individual nodes

2. Add a small `Transformer` class that is an `Interpreter` but provides a `transform` method that yields a GraphModule from interpreting the code

TODO:

- [ ] Write docstrings
- [ ] Write docs
- [ ] Look for more places this can be used

Differential Revision: [D25880330](https://our.internmc.facebook.com/intern/diff/D25880330)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Add Interpreter and Transformer"


1. Codify the interpreter pattern that was used, for ex, in [shape prop](https://github.com/pytorch/pytorch/blob/c3b4b2062748ed70e689f9f9ffe670e6fa20a071/torch/fx/experimental/shape_prop.py#L7). This is a pattern users will likely find repeatedly useful so it's good to have it first-class in the API. As seen in the test cases, this makes it incredibly easy to do simple analysis and transforms like swapping out individual nodes

2. Add a small `Transformer` class that is an `Interpreter` but provides a `transform` method that yields a GraphModule from interpreting the code

TODO:

- [ ] Write docstrings
- [ ] Write docs
- [ ] Look for more places this can be used

Differential Revision: [D25880330](https://our.internmc.facebook.com/intern/diff/D25880330)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Add Interpreter and Transformer"


1. Codify the interpreter pattern that was used, for ex, in [shape prop](https://github.com/pytorch/pytorch/blob/c3b4b2062748ed70e689f9f9ffe670e6fa20a071/torch/fx/experimental/shape_prop.py#L7). This is a pattern users will likely find repeatedly useful so it's good to have it first-class in the API. As seen in the test cases, this makes it incredibly easy to do simple analysis and transforms like swapping out individual nodes

2. Add a small `Transformer` class that is an `Interpreter` but provides a `transform` method that yields a GraphModule from interpreting the code

TODO:

- [ ] Write docstrings
- [ ] Write docs
- [ ] Look for more places this can be used

Differential Revision: [D25880330](https://our.internmc.facebook.com/intern/diff/D25880330)

[ghstack-poisoned]

Commit:
Update on "[FX] Add Interpreter and Transformer"


1. Codify the interpreter pattern that was used, for ex, in [shape prop](https://github.com/pytorch/pytorch/blob/c3b4b2062748ed70e689f9f9ffe670e6fa20a071/torch/fx/experimental/shape_prop.py#L7). This is a pattern users will likely find repeatedly useful so it's good to have it first-class in the API. As seen in the test cases, this makes it incredibly easy to do simple analysis and transforms like swapping out individual nodes

2. Add a small `Transformer` class that is an `Interpreter` but provides a `transform` method that yields a GraphModule from interpreting the code

TODO:

- [x] Write docstrings
- [ ] Write docs
- [ ] Look for more places this can be used

Differential Revision: [D25880330](https://our.internmc.facebook.com/intern/diff/D25880330)

[ghstack-poisoned]

Commit:
Update on "[FX] Add Interpreter and Transformer"


1. Codify the interpreter pattern that was used, for ex, in [shape prop](https://github.com/pytorch/pytorch/blob/c3b4b2062748ed70e689f9f9ffe670e6fa20a071/torch/fx/experimental/shape_prop.py#L7). This is a pattern users will likely find repeatedly useful so it's good to have it first-class in the API. As seen in the test cases, this makes it incredibly easy to do simple analysis and transforms like swapping out individual nodes

2. Add a small `Transformer` class that is an `Interpreter` but provides a `transform` method that yields a GraphModule from interpreting the code

![image](https://user-images.githubusercontent.com/4685384/105924691-a7b0b700-5ff3-11eb-8de8-7967d6b8ac69.png)
![image](https://user-images.githubusercontent.com/4685384/105924696-abdcd480-5ff3-11eb-8454-a2bc36c3248e.png)


TODO:

- [x] Write docstrings
- [ ] Write docs
- [ ] Look for more places this can be used

Differential Revision: [D25880330](https://our.internmc.facebook.com/intern/diff/D25880330)

[ghstack-poisoned]

Commit:
Update on "[FX] Add Interpreter and Transformer"


1. Codify the interpreter pattern that was used, for ex, in [shape prop](https://github.com/pytorch/pytorch/blob/c3b4b2062748ed70e689f9f9ffe670e6fa20a071/torch/fx/experimental/shape_prop.py#L7). This is a pattern users will likely find repeatedly useful so it's good to have it first-class in the API. As seen in the test cases, this makes it incredibly easy to do simple analysis and transforms like swapping out individual nodes

2. Add a small `Transformer` class that is an `Interpreter` but provides a `transform` method that yields a GraphModule from interpreting the code

![image](https://user-images.githubusercontent.com/4685384/105924691-a7b0b700-5ff3-11eb-8de8-7967d6b8ac69.png)
![image](https://user-images.githubusercontent.com/4685384/105924696-abdcd480-5ff3-11eb-8454-a2bc36c3248e.png)


TODO:

- [x] Write docstrings
- [ ] Write docs
- [ ] Look for more places this can be used

Differential Revision: [D25880330](https://our.internmc.facebook.com/intern/diff/D25880330)

[ghstack-poisoned]

zdevito commented on torch/fx/interpreter.py:
```diff
@@ -0,0 +1,353 @@
+        Returns:
+            Any: The argument value that was retrieved.
+        """
+        return next(self.args_iter)
```
This doesn't handle `*args` correctly.

zdevito commented on torch/fx/interpreter.py:
```diff
@@ -0,0 +1,353 @@
+        self.submodules = dict(self.module.named_modules())
+        self.env : Dict[Node, Any] = {}
+
+        self.clear_env_on_return : bool = True
```
Maybe run should just take an initial environment, and otherwise substitute a blank one?

zdevito reviewed (commented):
Looks pretty reasonable. I have a few suggestions to better decompose the components into reusable parts.

jamesr66a commented on docs/source/fx.rst:
```diff
@@ -322,3 +322,9 @@ API Reference
+.. autoclass:: torch.fx.Interpreter
+  :members:
+
+.. autoclass:: torch.fx.Transformer
```
I haven't heard any good suggestions for this. Is the concern that this is the same name as the `Transformer` architecture? In that case, I would argue that "Transformer" is a terrible name for that architecture and I don't think we should let them squat on an English language common noun

Commit:
Update on "[FX] Add Interpreter and Transformer"


1. Codify the interpreter pattern that was used, for ex, in [shape prop](https://github.com/pytorch/pytorch/blob/c3b4b2062748ed70e689f9f9ffe670e6fa20a071/torch/fx/experimental/shape_prop.py#L7). This is a pattern users will likely find repeatedly useful so it's good to have it first-class in the API. As seen in the test cases, this makes it incredibly easy to do simple analysis and transforms like swapping out individual nodes

2. Add a small `Transformer` class that is an `Interpreter` but provides a `transform` method that yields a GraphModule from interpreting the code

![image](https://user-images.githubusercontent.com/4685384/105924691-a7b0b700-5ff3-11eb-8de8-7967d6b8ac69.png)
![image](https://user-images.githubusercontent.com/4685384/105924696-abdcd480-5ff3-11eb-8454-a2bc36c3248e.png)


TODO:

- [x] Write docstrings
- [x] Write docs
- [ ] Look for more places this can be used

Differential Revision: [D25880330](https://our.internmc.facebook.com/intern/diff/D25880330)

[ghstack-poisoned]

Commit:
Update on "[FX] Add Interpreter and Transformer"


1. Codify the interpreter pattern that was used, for ex, in [shape prop](https://github.com/pytorch/pytorch/blob/c3b4b2062748ed70e689f9f9ffe670e6fa20a071/torch/fx/experimental/shape_prop.py#L7). This is a pattern users will likely find repeatedly useful so it's good to have it first-class in the API. As seen in the test cases, this makes it incredibly easy to do simple analysis and transforms like swapping out individual nodes

2. Add a small `Transformer` class that is an `Interpreter` but provides a `transform` method that yields a GraphModule from interpreting the code

![image](https://user-images.githubusercontent.com/4685384/105924691-a7b0b700-5ff3-11eb-8de8-7967d6b8ac69.png)
![image](https://user-images.githubusercontent.com/4685384/105924696-abdcd480-5ff3-11eb-8454-a2bc36c3248e.png)


TODO:

- [x] Write docstrings
- [x] Write docs
- [ ] Look for more places this can be used

Differential Revision: [D25880330](https://our.internmc.facebook.com/intern/diff/D25880330)

[ghstack-poisoned]

Commit:
Update on "[FX] Add Interpreter and Transformer"


1. Codify the interpreter pattern that was used, for ex, in [shape prop](https://github.com/pytorch/pytorch/blob/c3b4b2062748ed70e689f9f9ffe670e6fa20a071/torch/fx/experimental/shape_prop.py#L7). This is a pattern users will likely find repeatedly useful so it's good to have it first-class in the API. As seen in the test cases, this makes it incredibly easy to do simple analysis and transforms like swapping out individual nodes

2. Add a small `Transformer` class that is an `Interpreter` but provides a `transform` method that yields a GraphModule from interpreting the code

![image](https://user-images.githubusercontent.com/4685384/105924691-a7b0b700-5ff3-11eb-8de8-7967d6b8ac69.png)
![image](https://user-images.githubusercontent.com/4685384/105924696-abdcd480-5ff3-11eb-8454-a2bc36c3248e.png)


TODO:

- [x] Write docstrings
- [x] Write docs
- [ ] Look for more places this can be used

Differential Revision: [D25880330](https://our.internmc.facebook.com/intern/diff/D25880330)

[ghstack-poisoned]

Commit:
Update on "[FX] Add Interpreter and Transformer"


1. Codify the interpreter pattern that was used, for ex, in [shape prop](https://github.com/pytorch/pytorch/blob/c3b4b2062748ed70e689f9f9ffe670e6fa20a071/torch/fx/experimental/shape_prop.py#L7). This is a pattern users will likely find repeatedly useful so it's good to have it first-class in the API. As seen in the test cases, this makes it incredibly easy to do simple analysis and transforms like swapping out individual nodes

2. Add a small `Transformer` class that is an `Interpreter` but provides a `transform` method that yields a GraphModule from interpreting the code

![image](https://user-images.githubusercontent.com/4685384/105924691-a7b0b700-5ff3-11eb-8de8-7967d6b8ac69.png)
![image](https://user-images.githubusercontent.com/4685384/105924696-abdcd480-5ff3-11eb-8454-a2bc36c3248e.png)


TODO:

- [x] Write docstrings
- [x] Write docs
- [ ] Look for more places this can be used

Differential Revision: [D25880330](https://our.internmc.facebook.com/intern/diff/D25880330)

[ghstack-poisoned]
